Improve the td integration method in ptsampler

ga7g08 opened the pull request:
Previously this used a simple quadrature, but this introduced a bug
which tended to end in overestimates of the evidence. Replacing this
with numpy's trapz rule avoid this bug and improves the readability

@farr : Could you take a look at this and check I haven't misunderstood
the original code?

farr commented:
I have a couple of questions:

* Are you sure that your method is more accurate?  It will be if one of the
betas in the ladder is 0, but if you don't include a T = infinity, beta = 0
chain, then you will find that it is systematically off, since it ignores
the area between min(betas) and beta = 0.  That was the rationale behind
the original Riemann sum.

* What happened to the "-"?  You want to integrate from beta = 0 to beta =
1, but the betas are stored in the opposite order in the array, right?

On Fri, May 22, 2015 at 9:37 AM Gregory Ashton <notifications@github.com>
wrote:

> Previously this used a simple quadrature, but this introduced a bug
> which tended to end in overestimates of the evidence. Replacing this
> with numpy's trapz rule avoid this bug and improves the readability
>
> @farr <https://github.com/farr> : Could you take a look at this and check
> I haven't misunderstood
> the original code?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/dfm/emcee/pull/155
> Commit Summary
>
>    - Improve the td integration method in ptsampler
>
> File Changes
>
>    - *M* emcee/ptsampler.py
>    <https://github.com/dfm/emcee/pull/155/files#diff-0> (8)
>
> Patch Links:
>
>    - https://github.com/dfm/emcee/pull/155.patch
>    - https://github.com/dfm/emcee/pull/155.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dfm/emcee/pull/155>.
>

ga7g08 commented:
Firstly I am extremely embarrassed to admit that I didn't run the test properly before
pushing (I forgot to install the modified version before running the test) and hence 
dropped the minus. I've added that in now and perhaps before this is ever pulled in
I will  squash it into one sensible commit. 

Secondly thanks for explaining the rationale for the Riemann sum, I hadn't appreciated that
that was the intention. However, since numpy calculated the difference as `out[n] = a[n+1] - a[n]`, then the implementation is a right-Riemann sum and therefore is an overestimation since <loglike> is monotonically increasing. 

So it seems my method neglects the beta -> 0 part of the integration, while the right-Riemann sum over estimates the integral. Personally I feel that the user should be checking (or perhaps should be forced to check) the plot of beta againat lnlike to ensure they have reasonable values (e.g lnlike is sufficiently small for there smallest beta). In this case I prefer trapz since it is more readable, but the obvious alternative is a middle Riemann sum?

PS: I am wondering if the issue discussed in #131 may be solved here? I ran the `multimodal.py` example as given in @drphilmarshall commit and got 

```
Estimated log evidence =  -64.0249172494 +/- 152.159339428
Analytic log evidence =  -6.67931612501
```

PPS: If you want an example of the overestimation [here is a gist containing the minimal code](https://gist.github.com/ga7g08/2874c19a55f0a2452135) with results:

```
Actual ln(Z) = 93.4636629474
lnZ_m1=93.1525160422, err=0.311146905153
lnZ_m2=95.5127068853, err=2.04904393787
```

farr commented:
Actually, there is a bug in Marshall's multimodal.py, which skips the
normalisation branch of the if statement in the prior.  I uploaded a fixed
version, and transitioned to your trapz integration (you're right---it's
better to have a more accurate integral over the range than to try to
squeeze the last bit of integrand toward beta = 0) in PR #154.  Go ahead
and have a look.

On Fri, May 22, 2015 at 10:57 AM Gregory Ashton <notifications@github.com>
wrote:

> Firstly I am extremely embarrassed to admit that I didn't run the test
> properly before
> pushing (I forgot to install the modified version before running the test)
> and hence
> dropped the minus. I've added that in now and perhaps before this is ever
> pulled in
> I will squash it into one sensible commit.
>
> Secondly thanks for explaining the rationale for the Riemann sum, I hadn't
> appreciated that
> that was the intention. However, since numpy calculated the difference as out[n]
> = a[n+1] - a[n], then the implementation is a right-Riemann sum and
> therefore is an overestimation since is monotonically increasing.
>
> So it seems my method neglects the beta -> 0 part of the integration,
> while the right-Riemann sum over estimates the integral. Personally I feel
> that the user should be checking (or perhaps should be forced to check) the
> plot of beta againat lnlike to ensure they have reasonable values (e.g
> lnlike is sufficiently small for there smallest beta). In this case I
> prefer trapz since it is more readable, but the obvious alternative is a
> middle Riemann sum?
>
> PS: I am wondering if the issue discussed in #131
> <https://github.com/dfm/emcee/pull/131> may be solved here? I ran the
> multimodal.py example as given in @drphilmarshall
> <https://github.com/drphilmarshall> commit and got
>
> Estimated log evidence =  -64.0249172494 +/- 152.159339428
> Analytic log evidence =  -6.67931612501
>
> PPS: If you want an example of the overestimation here is a gist
> containing the minimal code
> <https://gist.github.com/ga7g08/2874c19a55f0a2452135> with results:
>
> Actual ln(Z) = 93.4636629474
> lnZ_m1=93.1525160422, err=0.311146905153
> lnZ_m2=95.5127068853, err=2.04904393787
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dfm/emcee/pull/155#issuecomment-104603099>.
>

ga7g08 commented:
Okay thanks, I added a note that I think you are missing a minus sign as I did ( I guess it has been a long day :) ) Should I close this then as it is no longer needed?

farr commented:
I think you can go ahead and close it, once you verify that #154 works for
you as it did for me.

On Fri, May 22, 2015 at 1:41 PM Gregory Ashton <notifications@github.com>
wrote:

> Okay thanks, I added a note that I think you are missing a minus sign as I
> did ( I guess it has been a long day :) ) Should I close this then as it is
> no longer needed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dfm/emcee/pull/155#issuecomment-104649959>.
>
